metrics.go

```diff
@@ -1,122 +1,173 @@
 package main
 
-import "github.com/prometheus/client_golang/prometheus"
+import (
+	"github.com/contentsquare/chproxy/config"
+	"github.com/prometheus/client_golang/prometheus"
+)
 
 var (
+	statusCodes                    *prometheus.CounterVec
+	requestSum                     *prometheus.CounterVec
+	requestSuccess                 *prometheus.CounterVec
+	limitExcess                    *prometheus.CounterVec
+	hostPenalties                  *prometheus.CounterVec
+	hostHealth                     *prometheus.GaugeVec
+	concurrentQueries              *prometheus.GaugeVec
+	requestQueueSize               *prometheus.GaugeVec
+	userQueueOverflow              *prometheus.CounterVec
+	clusterUserQueueOverflow       *prometheus.CounterVec
+	requestBodyBytes               *prometheus.CounterVec
+	responseBodyBytes              *prometheus.CounterVec
+	cacheHit                       *prometheus.CounterVec
+	cacheMiss                      *prometheus.CounterVec
+	cacheSize                      *prometheus.GaugeVec
+	cacheItems                     *prometheus.GaugeVec
+	requestDuration                *prometheus.SummaryVec
+	proxiedResponseDuration        *prometheus.SummaryVec
+	cachedResponseDuration         *prometheus.SummaryVec
+	canceledRequest                *prometheus.CounterVec
+	cacheHitFromConcurrentQueries  *prometheus.CounterVec
+	cacheMissFromConcurrentQueries *prometheus.CounterVec
+	killedRequests                 *prometheus.CounterVec
+	timeoutRequest                 *prometheus.CounterVec
+	configSuccess                  prometheus.Gauge
+	configSuccessTime              prometheus.Gauge
+	badRequest                     prometheus.Counter
+)
+
+func initMetrics(cfg *config.Config) {
+	namespace := getNamespace(cfg)
 	statusCodes = prometheus.NewCounterVec(
 		prometheus.CounterOpts{
-			Name: "status_codes_total",
-			Help: "Distribution by status codes",
+			Namespace: namespace,
+			Name:      "status_codes_total",
+			Help:      "Distribution by status codes",
 		},
```

Comment:
Not sure what will happen if the Namespace field is an empty string. Maybe the better option will be to implement a function that returns `prometheus.CounterOpts` . So, in case  `Namespace` is set in the global config, `Namespace` will include in the function return.

Reply:
when the Namespace field is an empty string, it will not append anything before metric name.